core-processor/src/ext.rs

```diff
@@ -670,6 +674,46 @@ impl EnvExt for Ext {
         Ok(self.context.message_context.current().id())
     }
 
+    fn pay_rent(&mut self, program_id: ProgramId, rent: u128) -> Result<u128, Self::Error> {
+        let block_count = u32::try_from(rent / self.context.rent_cost).unwrap_or(u32::MAX);
+
+        let old_paid_blocks = self
+            .context
+            .program_rents
+            .get(&program_id)
+            .copied()
+            .unwrap_or(0);
+
+        let (paid_blocks, blocks_to_pay) = match old_paid_blocks.overflowing_add(block_count) {
+            (count, false) => (count, block_count),
+            (_, true) => (u32::MAX, u32::MAX - old_paid_blocks),
+        };
+
+        if blocks_to_pay == 0 {
+            return Ok(rent);
+        }
+
+        let cost = self.context.rent_cost * u128::from(blocks_to_pay);
+        match self.context.value_counter.reduce(cost) {
+            ChargeResult::Enough => {
+                self.context
+                    .program_rents
+                    .entry(program_id)
+                    .and_modify(|rent_blocks| *rent_blocks = paid_blocks)
+                    .or_insert(block_count);
+            }
+            ChargeResult::NotEnough => {
+                return Err(ExecutionError::NotEnoughValueForRent {
```

Comment:
better to add info about blocks amount (I would still go with approach of amount of blocks)

Reply:
discussed: return the `rent`